keps/sig-api-machinery/3962-mutating-admission-policies/README.md

```diff
@@ -0,0 +1,1514 @@
+| list        | map        | See below                                                                |
+| list        | granular   | Not Supported                                                            | 
+
+
+List with "map" merge type:
+  - Filter `objects.filter(<list>, <keys-to-remove>)` could be used for the deletion
+  - For associatedList with multiple keys like example above, a directive field added could be used to indicate the deletion.
+```yaml
+mutations:
+  - mutaton: >
+      Object{
+        spec: Object.spec{
+                assocListField: [Object.spec.assocListField{
+                       keyField1: "key1",
+                       keyField2: "key2",
+                       _: optional.none()
+                }]
+        }
+}
+
+```
+
+For examples of removing item from List with Map filtered by a subfield:
+```yaml
+mutations:
+  - mutaton: >
+        Object{
+            spec: Object.spec{
+                containers: object.spec.containers.filter(c, c.envvar != "remove-this-container")
+                }
+            }
+        }
+
+```
+
+For granular list removal, a use case would be removing an item with a sub field named `remove-this-item`.
+```yaml
+mutations:
+  - mutation: >
+      Object{
+        spec: Object.spec{
+                granularList: object.spec.granularList.filter(c, c.subField != "remove-this-item")
+        }
+      }
+```
+
+##### Safety
+
+To ensure mutations are not "broken" by other mutations (overwritten, undone, or
+otherwise invalidated) and ensure the deterministic final state due to the random running order, \
+we provide an option to check if rerun certain mutation policy leads to object change.
+Ir also helps to ensure the mutation is written in a idempotent way in consideration of the random running ordering. 
+
+These validation checks will be declared using a `mutationValidationPolicy` field, which is an enum of the following values:
+- Fail - Replaying the mutation on the mutated object should result in an identical object,  if not, fail the request.
+- Warn - Replaying the mutation on the mutated object should result in an identical object,  if not, pass the request with a warning message.
+- Skip - Don't replay the mutation.
+
+For example:
+
+```yaml
+apiVersion: admissionregistration.k8s.io/v1alpha1
+kind: MutatingAdmissionPolicy
+metadata:
+  name: "sidecar-policy.example.com"
+spec:
+  # ...
+  mutations:
+      mutation: >
+        Object{
+          spec: Object.spec{
+            initContainers: [
+              Object.spec.initContainers{
+                name: params.name,
+                image: params.image,
+                args: params.args,
+                restartPolicy: params.restartPolicy
+                // ... other container fields injected here ...
+              }
+            ] + object.spec.initContainers
+          }
+        }
+  mutationValidationPolicy: Fail
+```
+If rerun the mutation policy caused object change, the request should be failed
```

Comment:
Will mutations always be re-run, or only when there are changes later in the chain?

I'm wondering what would happen if I set this to fail, and it works happily for a long time, but lets say, adds a new entry to a list every time it runs. As a user, I _may_ notice this, and say oops, I need to go fix that. If this check was always run, then the API server would have told me straight away

Reply:
We would love to have it already re-run but maybe don't fail the request if specified so. And rerun is mainly for checking/validating which will not apply the changes.

Reply:
Yeah makes sense, just thinking from an end user perspective, if it only re-runs sometimes, the behaviour of the fail policy may be a little cryptic